storage/namespace.go

```diff
@@ -547,12 +594,14 @@ func (n *dbNamespace) NeedsFlush(blockStart time.Time) bool {
 		if shard == nil {
 			continue
 		}
-		state := shard.FlushState(blockStart)
-		if state.Status == fileOpNotStarted {
-			return true
-		}
-		if state.Status == fileOpFailed && state.NumFailures < maxRetries {
-			return true
+		for _, blockStart := range blockStarts {
```

Comment:
Yeah that's a pretty awful cost to pay, fwiw the version running in prod does this too.

I think it's a simple fix to avoid doing this: We can additionally make `namespace.Flush()` check shard state before invoking `shard.Flush()`. NB: we need to leave the NeedsFlush method the same as the version proposed in the PR because the Flush|Cleanup manager shouldn't be operating a shard granularity. 

Cool?